Add range check for 'fd' argument of BIO_socket_wait()

DDvO opened the pull request:
as discussed in https://github.com/openssl/openssl/pull/11971#discussion_r431817997.

bernd-edlinger commented on crypto/bio/b_sock.c:
```diff
@@ -387,6 +387,8 @@ int BIO_socket_wait(int fd, int for_read, time_t max_time)
     struct timeval tv;
     time_t now;
 
+    if (fd < 0 || fd >= FD_SETSIZE)
```
Okay, but I'd like to see the `fd >= FD_SETSIZE` also in bio_wait, and defer to the default handling
in that case.

DDvO replied:
Good point - done.

On this occasion I've also renamed the `millisecondss` param of `BiO_wait()` to `fallback_milliseconds` for clarity.

levitte commented:
This looks good to me, but I'll defer to @bernd-edlinger to actually approve

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -796,31 +796,32 @@ static int bio_wait(BIO *bio, time_t max_time, unsigned int milliseconds)
         return 1;
```
I wonder why the condition above this line `if (max_time == 0) `
makes sense, and why the return value is not 0 (timeout) in this case.

DDvO replied:
As written in the documentation of `BIO_socket_wait()`, `BIO_wait()`, and `BIO_connect_retry()`, the meaning of `max_time == 0` is that no timeout is wanted (i.e., blocking behavior has been selected).
This, as well as the 'success' return in this case, is for convenient use for instance by `BIO_connect_retry()` and the HTTP client, where `max_time` has the same semantics.

bernd-edlinger commented on crypto/bio/b_sock.c:
```diff
@@ -387,6 +387,8 @@ int BIO_socket_wait(int fd, int for_read, time_t max_time)
+    if (fd < 0 || fd >= FD_SETSIZE)
+        return -1;
     if (max_time == 0)
         return 1;
```
likewise here,  what is so special with max_time==0 that this
returns 1, which means socket is readable?

DDvO replied:
See above.
As written in the code comments of `BIO_connect_retry()`, there are some strange situations where also in the blocking case `BIO_do_connect()` returns with a retry condition, and for uniformity with the non-blocking case (with non-zero timeout/max_time) it is convenient to return with 1 (success) of `max_time == 0`.

bernd-edlinger replied:
I do not understand what strange condition that might be.
BUT I think I don't like libcrypto calling into libssl (BIO_do_handshake)

bernd-edlinger replied:
sorry mixed it up with SSL_do_handshake :-)

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -796,31 +796,32 @@ static int bio_wait(BIO *bio, time_t max_time, unsigned int milliseconds)
-        if ((unsigned long)sec_diff < milliseconds / 1000)
-            milliseconds = (unsigned long)sec_diff * 1000;
+        if ((unsigned long)sec_diff < fallback_millisec / 1000)
+            fallback_millisec = (unsigned long)sec_diff * 1000;
```
why are you casting to unsigned long, when fallback_millisec is unsigned int ?

DDvO replied:
This is a leftover from an earlier version where the type of `(fallback_)millisec` was `long` and not needed any more. Removed.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -796,31 +796,32 @@ static int bio_wait(BIO *bio, time_t max_time, unsigned int milliseconds)
         return BIO_socket_wait(fd, BIO_should_read(bio), max_time);
 #endif
-    if (milliseconds > 1000) {
+    if (fallback_millisec > 1000) {
```
Strictly speaking this if (fallback_millisec > 1000) looks wrong.
Why is the due time not considered when fallback millisec is <= 1000 ?

DDvO replied:
Here we are in the fallback case where proper waiting until max_time or input is available is not possible.
The idea is that we need to poll but not too tightly (to prevent a busy loop).
So we wait typically for the given number of milliseconds (such that the granularity of the polling can be determined by the caller) but at most for one second, which is the resolution of `max_time`.
This is also (to some extent) mentioned in the doc:
```
If sockets are not available it succeeds after waiting at most given
B<fallback_milliseconds> in order to avoid a tight busy loop at the caller.
```

bernd-edlinger replied:
but the time out check is in the if bloc,
so it will never happen.

bernd-edlinger replied:
how I see this you wait for a small time,
try to read, that returns EAGAIN, wait more
in a loop, that while loop has a time out.
But Why is max_time only compared against time()
when the "small time" is > 1000 ???

bernd-edlinger replied:
And, sorry, for what is this function useful when OPENSSL_NO_SOCK is defined?

DDvO replied:
Good point - there was a bug.
And my above description was not right regarding waiting at most one second.
I've corrected the code and extended the commenting inside `bio_wait()` and the doc of `BIO_wait()`.
Hope things are clear and correct now.

DDvO replied:
> And, sorry, for what is this function useful when OPENSSL_NO_SOCK is defined?

As I wrote, the function is designed to be useful also in the case no sockets are available, where one needs to resort to polling. The caller can decide via `fallback_milliseconds` how tight the busy loop shall be.

bernd-edlinger replied:
as a user I would expect to exclude socket-related code when I configure with no-sock
WTF is the use case, where you would use this function instead of a simple usleep,
when it ignores the passed in bio ?

DDvO replied:
I've meanwhile extended the documentation and comments of both `BIO_wait()` and `BIO_connect_retry()`, which should make their inter-working and use more clear.

Note that not all BIOs are socket-based, and even blocking BIOs sometimes return early (for unclear reasons) and ask for retry. For the user all that should not matter.
For non-blocking behavior a timeout value may need to be respected.

Therefore `BIO_connect_retry()` uses `BIO_wait()` to implement this transparently, for any type of BIO.
For non-blocking and potentially even non-socket BIOs it polls every 100 milliseconds and sleeps in between, avoiding a busy loop.

Moreover, the pretty convenient `BIO_wait()` is used also in `crypto/http/http_client.c` to transparently wait for new input in between `BIO_read()` or `BIO_gets()`.

bernd-edlinger replied:
Hmm, but the BIO_connect_retry is only for sockets.

DDvO replied:
It originally was meant for use with sockets.
Yet there might be also BIOs not involving sockets (well, of course not those implemented by `bss_conn.c`).
And as mentioned the polling is needed for sure also for non-blocking socket-based BIOs.

bernd-edlinger replied:
Can you give more details for when we needed polling for non-blocking socket BIOs?

DDvO replied:
Connection attempts for non-blocking (socket-based or not) BIOs typically return rather quickly. If retry is requested this would lead to a busy loop re-doing the retry immediately over and over.
I am sure about this because before I implemented the wait/sleep mechanism my CPU load went to 100% until the request max time was reached (timeout).

bernd-edlinger replied:
you only tried localhost ?
for connections involving possibly non-reachable network addresses
things are much different.

DDvO replied:
I agree, as for localhost it usually does not help much waiting for connection establishment.
I definitely tried for non-local severs and remember unplugging and plugging my laptop's ethernet cable (with WiFi disable) for trying this out :smile:

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -786,41 +786,48 @@ void bio_cleanup(void)
+    if (fallback_millisec > 1000) {
+
+        if (sec_diff < fallback_millisec / 1000)
+            fallback_millisec = sec_diff * 1000;
```
I did not mean no typecast at all.
I just wanted to say that I would be more appropriate to do
```C
if (sec_diff < fallback_millisec / 1000)
    fallback_millisec = (unsigned int) sec_diff * 1000;
```

bernd-edlinger replied:
whee, so make that
```C
if ((unsigned long)sec_diff < fallback_millisec / 1000)
    fallback_millisec = (unsigned int) sec_diff * 1000;
```

DDvO replied:
Done

t8m commented:
The AppVeyor failure is relevant.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -856,21 +863,21 @@ int BIO_connect_retry(BIO *bio, int timeout)
-             * when non-blocking, BIO_do_handshake() timed out early
-             * with rv == -1 and get_last_sys_error() == 0
+            /*-
+             * If blocking, BIO_do_handshake() timed out despite blocking BIO.
```
This is probably a misunderstanding how a blocking connect works.
If the server or a firewall does not send RST packets the connect
will time-out after 1-2 minutes.  So a connection error may be
immediately due to a RST packet, or after a timeout due to no SYN-ACK packet.
In both cases the next address in this list should be tried,
I wonder if the BIO_reset resets that address.

DDvO replied:
Thanks for that explanation.
AFAICS from `bio_conn.c` BIO_reset() resets the list of addresses to try, so does too much, right?

bernd-edlinger replied:
Could be possible

DDvO replied:
So I've just added for now the following comment:
```
            /*
             * TODO BIO_reset() maybe needs to be replaced since it likely
             * does too much, reseting the list of IP addresses to try
             */
```

DDvO replied:
I've meanwhile taken the time to re-do my connection error experiments and made adaptations in this PR accordingly.
* For non-blocking connects the first call to `BIO_do_connect()` always returns with "BIO should retry" and `errno=115` (Operation now in progress) and obviously needs to be retried, which works fine.
* For local ports (e.g., 127.0.0.1 or localhost:8000) that no service listens to there is an immediate "Connection refused" error, which I'd say is fine and there is likely no need to retry connecting until timeout.
* If there is no route to a non-local addresses (e.g., 1.2.3.4:8000 or test.com:8000) there is an immediate "No route to host" error, which I'd say is fine and there is likely no need to retry connecting until timeout.
* For unreachable non-local addresses (e.g., 1.2.3.4:8000 or test.com:8000) an internal timeout occurs after 30 seconds and `get_last_socket_error()` or `BIO_sock_error()` yields `110` (Connection timed out), where a blocking connect sets `errno=110` while a non-blocking connect for some reason retains `errno=115. In both blocking and non-blocking case `BIO_connect_retry()` should retry (including retry of any applicable IP addresses).
If the retry is done directly I cannot reproduce a crash, but a connect error is thrown again immediately. Therefore `ERR_clear_error(); (void)BIO_reset(bio);` are really necessary.
For enabling `BIO_connect_retry()` to correctly detecting these internal timeout (`ETIMEDOUT`) situations I've added for both the blocking and non-blocking case a differentiation of the error reason thrown: `BIO_R_CONNECT_TIMEOUT` and `BIO_R_NBIO_CONNECT_TIMEOUT`, respectively.
* For DNS trouble on non-local addresses (e.g., 1.2.3.4:8000 or test.com:8000) an internal timeout occurs after 40 seconds and `BIOerr(BIO_F_BIO_LOOKUP_EX, ERR_R_SYS_LIB)` is thrown and `get_last_sys_error()` yields `EAGAIN` (resource temporarily unavailable) and `BIO_connect_retry()` should retry.

The adaptations lead to major cleanup of `BIO_connect_retry()`.

bernd-edlinger replied:
> while a non-blocking connect for some reason retains errno=115.

I think that is basically what I claim that the errno is undefined when
the non-blocking socket connect fails:
```C
           i = BIO_sock_error(b->num);
            if (i != 0) {
                BIO_clear_retry_flags(b);
                if ((c->addr_iter = BIO_ADDRINFO_next(c->addr_iter)) != NULL) {
                    /*
                     * if there are more addresses to try, do that first
                     */
                    BIO_closesocket(b->num);
                    c->state = BIO_CONN_S_CREATE_SOCKET;
                    ERR_clear_error();
                    break;
                }
                ERR_raise_data(ERR_LIB_SYS, i,
                               "calling connect(%s, %s)",
                                c->param_hostname, c->param_service);
                BIOerr(BIO_F_CONN_STATE, BIO_R_NBIO_CONNECT_ERROR);
                ret = 0;
                goto exit_loop;
```
does nowere set errno, ETIMEDOUT
while `i = BIO_sock_error(b->num)` should return ETIMEDOUT,
so errno is still 115 EINPROGRESS.

DDvO replied:
Yes, so I've meanwhile made sure that `BIO_do_retry()` is independent of `errno`,
which in my view is cleaner anyway.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -856,21 +863,21 @@ int BIO_connect_retry(BIO *bio, int timeout)
-             * unless using BIO_reset(), blocking next connect() may crash and
-             * non-blocking next BIO_do_handshake() will fail
+             * Unless using BIO_reset(), blocking next connect() may crash and
+             * non-blocking next BIO_do_handshake() would fail.
```
I am a bit unsure if that is a symptom of a deeper problem.
Do you have some details what "crash" may have happened here?

DDvO replied:
Unfortunately I don't recall the details.
It was about two years back when I had trouble implementing a stable HTTP connect
and asked the OpenSSL gurus for help (as far as I remember on the mailing list).
As I did not receive it I experimented until I found a working solution, as given here.

bernd-edlinger replied:
I wouldn't block this PR for it, as it is just a range check, but this
explanation given in this comment raises more questions than it answers.
So probably need more investigations.

DDvO replied:
Meanwhile I found the email exchange I referred to 3 days ago: 
https://mta.openssl.org/pipermail/openssl-dev/2017-September/009697.html

DDvO replied:
The comments that - justifiably - raised questions are gone meanwhile,
due to the re-structuring of the retry checks that I did yesterday.
Are you fine with the new version?

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -856,21 +863,21 @@ int BIO_connect_retry(BIO *bio, int timeout)
 
```
I wonder if I look at this code above
```C
    if (!blocking)
        BIO_set_nbio(bio, 1);
```

is that garanteed that the bio is always in blocking mode so
wouldn't it be safer to call `BIO_set_nbio(bio, !blocking);` here ?

DDvO replied:
Good question. I had inferred that from programming examples.
Looking at the documentation of `BIO_set_nbio()` confirms this:
```
Blocking I/O is the default.
```

bernd-edlinger replied:
Yes, unless the BIO is initialized otherwise, and when you say,
the blocking sockets return with BIO_should_retry that makes me
curious, what triggers this behavior.

DDvO replied:
Would be great if you anyone else with more network programming experience than me could investigate into that. Shall we open an issue?

bernd-edlinger replied:
If you can setup something reproducible, yes, why not.
Currently it looks foggy to me.

DDvO replied:
All right, then I won't open an issue now because I currently cannot reproduce the crash,
and I think the newly added TODO should be sufficient for now.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -856,21 +863,21 @@ int BIO_connect_retry(BIO *bio, int timeout)
         } else if (BIO_should_retry(bio)) {
             /* will not actually wait if timeout == 0 (i.e., blocking BIO) */
-            rv = bio_wait(bio, max_time, 100 /* milliseconds */);
+            rv = bio_wait(bio, max_time, 100 /* milliseconds to sleep */);
```
I think you mentioned that this is sometimes called for non-blocking bio,
is that right?

DDvO replied:
Yes, this is meant for non-blocking BIOs.
As written in the comment it `bio_wait()` has no effect in the blocking case (timeout == 0).

bernd-edlinger replied:
I wonder why a blocking BIO should set BIO_should_retry, that is unexpected.

DDvO replied:
Right, `BIO_should_retry()` should be needed only for the non-blocking case.
I've meanwhile added a respective comment.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -856,21 +863,21 @@ int BIO_connect_retry(BIO *bio, int timeout)
 
     if (rv <= 0) {
         if (get_last_sys_error() == ETIMEDOUT) {
```
I would be surprised if that does not reset the errno,
the last call `i = BIO_sock_error(b->num);`
is not an error, so should not *change* errno,
in the end I believe using errno in this way uses undefined data?

DDvO replied:
Originally I had used `errno` but due to code review I replaced this by `get_last_sys_error()`. 
According to `e_os.h` it (usually) does not reset the errno:
```
# define get_last_sys_error()    errno
```

bernd-edlinger replied:
Yes, of course but errno is totally volatile, and the
value is *undefined* at least when the 
BIO_CONN_S_BLOCKED_CONNECT code path is
executed.
see `SYSerr(SYS_F_CONNECT, i);`

bernd-edlinger replied:
For instance errno is undefined when the error
is retrieved from getsockopt(s, SOL_SOCKET, SO_ERROR,...)

https://github.com/openssl/openssl/blob/082c041b4233b17b80129d4ac6b33a28014442b0/crypto/bio/bss_conn.c#L190-L193

DDvO replied:
This issue has been solved meanwhile, by avoiding the use of `errno`, as just mentioned above.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -786,41 +786,48 @@ void bio_cleanup(void)
+    if (polling_millisec > 1000) {
+
+        if ((unsigned long)sec_diff < polling_millisec / 1000)
+            polling_millisec = (unsigned int)sec_diff * 1000;
```
This does not work right, if polling_mllisec > 1000 
this will wait 1 seconds, until sec_diff == 0, then
since sec_diff < polling_millisec / 1000
set polling_millisec = 0,
then it will skip the wait below and create 100% CPU load
for the last second....

bernd-edlinger replied:
I think this is not completely right, this should not be skipped if polling_millisec <= 1000,
since sec_diff can be 0, but you do not sleep(0)

DDvO replied:
No, if `polling_millisec is > 1000` then the function determines `max(polling_millisec, 1000*sec_def)` and waits accordingly, as mentioned in the comment above that code.

bernd-edlinger replied:
but in last second of the timeout, 1000*sec_def is null, and then you have a busy loop.
jut for the last second, but I think that is not what you indend.

bernd-edlinger replied:
And actually this computes `polling_millisec = MIN(polling_millisec, 1000*sec_def)` 
not MAX. so it can be much smaller than polling_millisec.

DDvO replied:
Oops, I did actually mean: min.

DDvO replied:
If polling_millisec < 1000 the function does not wait zero time even if sec_diff is 0,
but for the intended nap period.

bernd-edlinger replied:
Yes, that case is of course more reasonable.
You wouldn't get that sort of problem if you
consider sec_diff = 0 as timeout.

DDvO replied:
Yes, but as documented I'd like to give the connect a last chance in case the remaining time before timeout is less than the resolution of the timeout value, which is one second.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -848,29 +857,20 @@ int BIO_connect_retry(BIO *bio, int timeout)
-            /* will not actually wait if timeout == 0 (i.e., blocking BIO) */
-            rv = bio_wait(bio, max_time, 100 /* milliseconds */);
+        if (BIO_should_retry(bio)
+                || get_last_sys_error() == EAGAIN /* Temporary gai error */) {
```
What is a "gai" error?
and if we end up here with EAGAIN and !BIO_should_retry,
that looks like a bug, so why not simply fix the bug?

DDvO replied:
By "gai" error I meant the following `getaddrinfo()` error: Temporary failure in name resolution.
I've improved the commenting and the handling of this case in my most recent push.
This is all I can do today; anything else needs to wait until tomorrow.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -848,29 +857,20 @@ int BIO_connect_retry(BIO *bio, int timeout)
+    if (polling_milliseconds < 0)
+        polling_milliseconds = 100;
     if (!blocking)
         BIO_set_nbio(bio, 1);
```
I'd suggest mak this BIO_set_nbio(bio, !blocking);
You can't be sure what the internal state of this bio is?

bernd-edlinger replied:
I think in case of blocking mode this should just call BIO_do_connect
there is no need for any retires in that case.
What exactly is the motivation for the non-blocking mode here,
put a time-out how long the connect may take?

DDvO replied:
> I'd suggest mak this BIO_set_nbio(bio, !blocking);
> You can't be sure what the internal state of this bio is?

Done.

DDvO replied:
> I think in case of blocking mode this should just call BIO_do_connect
> there is no need for any retires in that case.
> What exactly is the motivation for the non-blocking mode here,
> put a time-out how long the connect may take?

The retry is still needed in the non-blocking case, as meanwhile made clear in the revised code and its comments:
Even for blocking connects non-local connection attempts may return a premature error in case there a system-internal timeout occurs (e.g., when the intended host is currently not reachable) and in case a domain name resolution fails temporarily.
I've confirmed these situation in my yesterday's experiments.

bernd-edlinger replied:
> Done.

but not pushed :-)

> The retry is still needed 

So you re-try the connect even if it is a hard failure.
Now I understand at least what the code should do...

DDvO replied:
Oops, that commit went wrong. Now it's in and pushed :)

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -848,29 +857,20 @@ int BIO_connect_retry(BIO *bio, int timeout)
-    rv = BIO_do_handshake(bio); /* This indirectly calls ERR_clear_error(); */
+ retry: /* it does not always help to set SSL_MODE_AUTO_RETRY */
+    clear_sys_error();
+    rv = BIO_do_connect(bio); /* This indirectly calls ERR_clear_error(); */
```
looking at It bss_conn.c it don't look like always calling ERR_clear_error,
I only see that happening when the next address is tried (blocking mode)
or when a non-blocking connect is started.

DDvO replied:
Correct. Yet my comment was not meant as an assertion that `ERR_clear_error()` is called in any case but as a warning that it may be called, such that any previous error queue contents may be lost.
I've meanwhile made this clear, also in the documentation, in my today's fixup commit.

bernd-edlinger commented:
@DDvO I just spotted another issue in the bss_conn.c file:

https://github.com/openssl/openssl/blob/60d3b5b9ffb8c1273af0cc0338ec1c98f464f4b1/crypto/bio/bss_conn.c#L446-L457

You see case 1 is  BIO_set_conn_port, this should really be a comment here.
but it calls strdup and does not check the result succeeds
And case 2 is BIO_set_conn_address, here is a resource leak since the original
param_hostname and param_service need to be free'd.
Moreover the if(ret) is useless, since ret is just initialized to 1.
But also here is no checking if the operation succeeds.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -786,41 +786,48 @@ void bio_cleanup(void)
 
 #ifndef OPENSSL_NO_SOCK
-    if (BIO_get_fd(bio, &fd) > 0)
+    if (BIO_get_fd(bio, &fd) > 0 && fd < FD_SETSIZE)
```
Actually the strategy to fall back to polling can
work for read/write but wont work for connect.
Reason is that in the non-blocking case, you have
to wait until the socket is in writable state, but when
the select can't be used for that, you can't re-try
the connect, but have to use getsockopt(s, SOL_SOCKET, SO_ERROR, ...)
and that returns != 0 if an error happened,
but 0 when the connect succeeds, but also when the connect is
not yet completed.  You can assume the connect succeeds,
and continue with send/receive data which will fail, but we dont
have the data to send, and instead want to try the next
address if the previous one fails.

bernd-edlinger replied:
It would make the bio_wait function more useful if it tries
to use the poll(2) API if it is available, and just fail if it is
not available.  I would recommend not to fall back the sleep
and also not to expose this API at all when NO_SOCK is in effect.

DDvO replied:
> Actually the strategy to fall back to polling can
> work for read/write but wont work for connect.
> Reason is that in the non-blocking case, you have
> to wait until the socket is in writable state, but when
> the select can't be used for that, you can't re-try
> the connect, but have to use getsockopt(s, SOL_SOCKET, SO_ERROR, ...)
> and that returns != 0 if an error happened,
> but 0 when the connect succeeds, but also when the connect is
> not yet completed. You can assume the connect succeeds,
> and continue with send/receive data which will fail, but we dont
> have the data to send, and instead want to try the next
> address if the previous one fails.

> It would make the bio_wait function more useful if it tries
> to use the poll(2) API if it is available, and just fail if it is
> not available. I would recommend not to fall back the sleep
> and also not to expose this API at all when NO_SOCK is in effect.

I've tried several times to understand what exactly you meant here.
When resorting in the non-blocking case to polling then this is because the BIO is not socket-based and then one likely cannot use `getsockopt()` nor the poll API since they work on a socket.
For non-socket based connects it should still make sense to nap and poll again on a failed connect - AFAICS you don't have other chances than doing that (or simply failing potentially prematurely).

Please note also that `BIO_wait()` is also used for read/write, and at least there we agree that falling back to polling does make sense.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -848,35 +857,44 @@ int BIO_connect_retry(BIO *bio, int timeout)
+                do_retry = 1;
+            }
+            else if (reason == ERR_R_SYS_LIB && get_last_sys_error()
+                     == EAGAIN /* resource temporarily unavailable */) {
```
I personally, would stop this loop when the address resolution fails.
Technically it is problemati since when we come here, lots of code has run and could clobber errno,
would it be possible to get the original errno from there:
```C
       switch ((gai_ret = getaddrinfo(host, service, &hints, res))) {
# ifdef EAI_SYSTEM
        case EAI_SYSTEM:
            ERR_raise_data(ERR_LIB_SYS, get_last_socket_error(),
                           "calling getaddrinfo()");
            BIOerr(BIO_F_BIO_LOOKUP_EX, ERR_R_SYS_LIB);
```

DDvO replied:
Meanwhile I see that caling `get_last_socket_error()` is essentially the same as reading `errno`.

In my yesterday's experiments I confirmed that the (previous and current) implementation works fine:
If I use `BIO_do_connect_retry()` in blocking mode or with a sufficiently large timeout value and at first the DNS does not work but after some retries it starts working, the function gets to learn this fact in that way, and the connection succeeds.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -786,53 +786,62 @@ void bio_cleanup(void)
-        if ((unsigned long)sec_diff < milliseconds / 1000)
-            milliseconds = (unsigned long)sec_diff * 1000;
+    sec_diff = (long)(max_time - time(NULL)); /* might overflow */
+    if (sec_diff < 0)
```
previously this was if (sec_diff <= 0) was this change intentional?

DDvO replied:
Yes, because if `sec_diff == 0` the calculation of the minimum would go wrong and the sec_diff resolution insufficient if `polling_millisec` < 1000.

bernd-edlinger replied:
I don't understand your reasoning, as you say `if sec_diff == 0 the calculation of the minimum would go wrong`
I think you never want a sleep of 0 ms.  The easiest fix would be to consider sec_diff == 0 as time out.

DDvO replied:
By `would go wrong` if `sec_diff == 0` I meant that in this case `min(sec_diff * 1000, nap_milliseconds)` would be 0 ms, and this is what the code was intended to prevent. 
But the calculation was really buggy - I've just revised it.
Thanks for having a close look at that!

bernd-edlinger commented on crypto/bio/bss_conn.c:
```diff
@@ -202,15 +198,27 @@ static int conn_state(BIO *b, BIO_CONNECT *c)
 
         case BIO_CONN_S_CONNECT_ERROR:
-            BIOerr(BIO_F_CONN_STATE, BIO_R_CONNECT_ERROR);
+            i = get_last_socket_error();
```
the callback is called, and could do anything with the errno.

DDvO replied:
Sure a callback could do all kinds of things, but it should mess with `errno` only if there is good reason.
And how else would you suggest that one can detect the `getaddrinfo: Temporary failure in name resolution` situation?

bernd-edlinger replied:
Yes, the getaddrinfo is something I hadn't considered before.
I don't really know, but wouldn't it be possible to get the 
errno value back what was entered to the error queue, right
when the getaddrinfo failed?

But I don't fully agree that this change is an improvement,
the difference when a connect returns TIMEOUT and when
it returns a different error, is not very big.  Just an lost
packet for instance.  But if a server is not online, you can
also try again it might just starting.
For the do_connect_retry it would be okay to retry as long
as the time is over.  I had not considered that possibility
but that can be done, but then I would just do that.

DDvO replied:
> Yes, the getaddrinfo is something I hadn't considered before.
> I don't really know, but wouldn't it be possible to get the
> errno value back what was entered to the error queue, right
> when the getaddrinfo failed?

Indeed it would be possible and likely more safe to consult `errno` just after `getaddrinfo` failed, and then add a to-the-point entry/reason to the error queue

> But I don't fully agree that this change is an improvement,
> the difference when a connect returns TIMEOUT and when
> it returns a different error, is not very big. Just an lost
> packet for instance. But if a server is not online, you can
> also try again it might just starting.
> For the do_connect_retry it would be okay to retry as long
> as the time is over. I had not considered that possibility
> but that can be done, but then I would just do that.

You mean: as long the time is *not* over.
Good point, so we could at least to some extent lump timeout and error.
For some cases (`ETIMEDOUT`) I found that we need to call `BIO_reset()` to prevent immediate failure of the retry, but maybe such a call does not hurt also for all other error cases? Then much of the distinctions I added could be removed.

bernd-edlinger replied:
> Indeed it would be possible and likely more safe to consult errno just after getaddrinfo failed, and then add a to-the-point entry/reason to the error queue

I really don't know but isn't the errno here entered into the error queue?
https://github.com/openssl/openssl/blob/a5a87011baeef71c86938a2bae54f89fbe99e5dc/crypto/bio/b_addr.c#L705-L708

It would feel much safer to get the errno back from the error queue,
but if that is not possible, if the DNS says no this name is not known *now*
you could as well try again after the short nap.

It would this connect_retry less surprising when there function simply retries
what it is doing until that timeout is over or it succeeds then it leaves the loop.

>  I found that we need to call BIO_reset() 

I think I begin to understand what this does.  in the case when
the connect fails, the socket is *not* closed, just the *connect*
failed.  and the state stays at BIO_CONN_S_CONNECT_ERROR.
Without that reset the state machine will not continue with
BIO_CONN_S_BEFORE.

Yes, if that is the correct behaviour for your application you just
can reset the socket and try to connect until that time is over.
I just think that this may be a parameter, since some clients,
may just want to connect once, and let the user decide what to
do next.

DDvO replied:
> I really don't know but isn't the errno here entered into the error queue?

Yes, it is entered by `ERR_raise_data(ERR_LIB_SYS, get_last_socket_error(), "calling getaddrinfo()");` but then it is buried by `BIOerr(BIO_F_BIO_LOOKUP_EX, ERR_R_SYS_LIB);`.
Meanwhile `BIO_do_connect_retry()`, as you suggested, does not any more depend on `errno`.

DDvO replied:
> Yes, if that is the correct behaviour for your application you just
can reset the socket and try to connect until that time is over.

This is meanwhile implemented.

DDvO replied:
> I just think that this may be a parameter, since some clients,
may just want to connect once, and let the user decide what to
do next.

I like the idea to offer this option, although in this case the function degenerates to a normal `BIO_do_connect() call`.
Nevertheless I have implemented it. I did so not as an extra parameter but in case `timeout < 0`, which so far had no real use for `BIO_do_connect_retry()`.
This adds to uniformity and even simplifies the implementation.

bernd-edlinger replied:
Yes.  I think that is getting better now.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -848,35 +857,44 @@ int BIO_connect_retry(BIO *bio, int timeout)
+
+        if (ERR_GET_LIB(err) == ERR_LIB_BIO) {
+            if (reason == BIO_R_NBIO_CONNECT_TIMEOUT
+                    || reason == BIO_R_CONNECT_TIMEOUT) {
```
With the BIO_R_CONNECT_TIMEOUT this is really an endless loop.
the blocking connect is only tried when timeout == 0
I would also not differentiate between NBIO_CONNECT_TIMEOUT and NBIO_CONNECT_ERROR,
if the server does not listen, you can get a CONNECT_ERROR and you can of course
try that connection again after a short nap.

DDvO replied:
> With the BIO_R_CONNECT_TIMEOUT this is really an endless loop.
the blocking connect is only tried when timeout == 0

Indeed it could result in an endless loop (yet not in a busy loop, due to `BIO_wait`),
but if the user sets `timeout == 0` this is exactly what he wants: keep trying to connect as long as whatever needed (even infinitely) as long as the connection cannot be established.

> I would also not differentiate between NBIO_CONNECT_TIMEOUT and NBIO_CONNECT_ERROR,
if the server does not listen, you can get a CONNECT_ERROR and you can of course
try that connection again after a short nap.

Interesting thought. Yet I would assume that `NBIO_CONNECT_ERROR` is something fatal, while `NBIO_CONNECT_TIMEOUT` according to my experiments is a non-fatal error just indicating that within the strange 30 s internal grace period no connection could be established, so when the user is patient enough it makes sense to retry.

bernd-edlinger replied:
It is your function.

But again, I did not assume that timeout=0 means forever, I did just assume,
try the connection, to all addresses that match to the request exactly once.

For the blocking connect, you get only the error code of the last connect
attempt, the first few could have a CONNECT_ERROR and then a few
CONNECT_TIMEOUT and then it is an endless loop.
But if the first connect has a CONNECT_TIMEOUT and then a CONNECT_ERROR,
then you stop the loop.

bernd-edlinger replied:
Same in the non-blocking connect.
You get here only when the *last* address that getaddrinfo returned
had a TIMEOUT.  That looks bogus to take that as an indication
when to terminate the loop.

DDvO replied:
You just convinced me that it is more consistent - and also simpler of course - not to distinguish the timeout and error cases. 
I'll give that a try...

DDvO replied:
So I've unified the handling of  CONNECT_TIMEOUT and CONNECT_ERROR,
which is most likely an improvement, at least in terms of consistency.

It turns out that this does not yet cover the system error cases, in particular `EAGAIN` of `getaddrinfo` (Temporary failure in name resolution), so I've added the `ERR_R_SYS_LIB` case to cover all system errors, assuming that it makes sense to retry on any of them.

Experiments after these changes went fine. 
So it looks like we really can treat all non-fatal errors like internal  timeouts: reset the BIO, clear the error queue, and retry.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -848,35 +858,42 @@ int BIO_connect_retry(BIO *bio, int timeout)
 
- retry: /* it does not help here to set SSL_MODE_AUTO_RETRY */
-    rv = BIO_do_handshake(bio); /* This indirectly calls ERR_clear_error(); */
+ retry: /* it does not always help to set SSL_MODE_AUTO_RETRY */
```
The comment mentions SSL_MODE_AUTO_RETRY
This error handling is for BIO_TYPE_CONNECT,
not for BIO_TYPE_SSL, right?

DDvO replied:
I also think that SSL_MODE_AUTO_RETRY is specifically for SSL-type BIOs, and this is one of the reasons why in general it is not sufficient.

bernd-edlinger replied:
I just want to make sure this comment is not understood that this logic here is
prepared to handle BIO_TYPE_SSL, those have other special wait conditions,
for instance BIO_get_retry_reason() == BIO_RR_SSL_X509_LOOKUP

DDvO replied:
Do you have a suggestion how to phrase the comment better,
or should I simply drop it because it may give more confusion than explanation?

bernd-edlinger replied:
Yes please drop the comment.

DDvO replied:
Done

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -786,55 +786,65 @@ void bio_cleanup(void)
+        return 0; /* clearly timeout */
+
+    /* now take a nap at most the given number of milliseconds */
+    if (nap_milliseconds > 1000) {
```
This condition is redundant, if nap_milliseconds < 1000 the following
condition is sec_diff < 0 so never satisfied.
if nap_milliseconds == 1000 the following condition is only satisfied,
if sec_diff < 1, but this changes 1000ms to 0ms delay.
That is not what you want.

bernd-edlinger replied:
...and it does not happen when nap_milliseconds = 999, then you have
a sleep 999, in the last second.

DDvO replied:
I've revised the calculation. Does it make sense now?

bernd-edlinger replied:
yes.

bernd-edlinger commented on crypto/bio/bss_conn.c:
```diff
@@ -202,15 +198,27 @@ static int conn_state(BIO *b, BIO_CONNECT *c)
             break;
 
         case BIO_CONN_S_CONNECT_ERROR:
-            BIOerr(BIO_F_CONN_STATE, BIO_R_CONNECT_ERROR);
```
please drop this change.

DDvO replied:
I dropped the differentiation w.r.t. `ETIMEDOUT`, but the move of
```
                   ERR_raise_data(ERR_LIB_SYS, get_last_socket_error(),
                                  "calling connect(%s, %s)",
                                   c->param_hostname, c->param_service);
```
to the new code location is for consistency with the non-blocking case,
and the addition of the guard `i = get_last_socket_error(); if (i != 0)` not only is for consistency but likely also a necessary safeguard. At least for some (good?) reason this guard is present for the non-blocking case.

bernd-edlinger replied:
maybe I missed something, but you should never move the errno away from the
place where the error happened, and especially not call a callback in between.

DDvO replied:
All right, I've reverted this.
Likely the guard `i = get_last_socket_error(); if (i != 0)` is not critical.

bernd-edlinger commented on crypto/bio/bss_conn.c:
```diff
@@ -202,15 +198,27 @@ static int conn_state(BIO *b, BIO_CONNECT *c)
                                "calling connect(%s, %s)",
                                 c->param_hostname, c->param_service);
-                BIOerr(BIO_F_CONN_STATE, BIO_R_NBIO_CONNECT_ERROR);
+                if (i == ETIMEDOUT)
```
please drop this change.

DDvO replied:
Done. Also for the blocking case.

bernd-edlinger commented on crypto/bio/bss_conn.c:
```diff
@@ -452,19 +461,26 @@ static long conn_ctrl(BIO *b, int cmd, long num, void *ptr)
+                char *service = BIO_ADDR_service_string(addr, 1);
+
+                ret = host != NULL && service != NULL;
                 if (ret) {
```
remove this if, ret is always 1 here.

bernd-edlinger replied:
oops. disregard the previous comment.
instead you need to add an else block
and free the host and service in case only
one was NULL.

DDvO replied:
Ah yes, good that you reminded me of that case. Done.

bernd-edlinger commented on crypto/bio/bss_conn.c:
```diff
@@ -438,12 +446,13 @@ static long conn_ctrl(BIO *b, int cmd, long num, void *ptr)
         if (ptr != NULL) {
             b->init = 1;
-            if (num == 0) {
+            if (num == 0) { /* BIO_set_conn_hostname */
```
Thanks for the comments.

This part would be good to have in the 1.1.1 branch.

DDvO replied:
Argh, I just did this backport and by mistake directly pushed it to `OpenSSL_1_1_1-stable` :(
See #12034.

DDvO replied:
The actual backport PR is now in #12036.

DDvO commented:
> @DDvO I just spotted another issue in the bss_conn.c file:
> 
> https://github.com/openssl/openssl/blob/60d3b5b9ffb8c1273af0cc0338ec1c98f464f4b1/crypto/bio/bss_conn.c#L446-L457
> 
> You see case 1 is BIO_set_conn_port, this should really be a comment here.
> but it calls strdup and does not check the result succeeds
> And case 2 is BIO_set_conn_address, here is a resource leak since the original
> param_hostname and param_service need to be free'd.
> Moreover the if(ret) is useless, since ret is just initialized to 1.
> But also here is no checking if the operation succeeds.

Although these issues are rather far from the topic of this PR I've fixed them along with it, in commit 76d9cc01cdbed5eba472c2983cdca7a4c535528e.

bernd-edlinger commented on crypto/bio/bss_conn.c:
```diff
@@ -174,9 +173,6 @@ static int conn_state(BIO *b, BIO_CONNECT *c)
                 } else {
-                    ERR_raise_data(ERR_LIB_SYS, get_last_socket_error(),
-                                   "calling connect(%s, %s)",
-                                    c->param_hostname, c->param_service);
```
please undo this change.

DDvO replied:
Done

bernd-edlinger commented on crypto/bio/bss_conn.c:
```diff
@@ -8,7 +8,6 @@
  */
 
 #include <stdio.h>
-#include <errno.h>
```
maybe undo this too.

DDvO replied:
Done (though the #include is not really needed, 
but `get_last_socket_error()` indirectly uses `errno`).

bernd-edlinger replied:
I wear special roentgen glasses that look throu macros :-)

bernd-edlinger commented on crypto/bio/bss_conn.c:
```diff
@@ -210,6 +206,12 @@ static int conn_state(BIO *b, BIO_CONNECT *c)
             break;
 
         case BIO_CONN_S_CONNECT_ERROR:
+            i = get_last_socket_error();
```
please revert this change, reason that errno is likely no preseved over the callback.

DDvO replied:
Good point, done.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -843,40 +854,47 @@ int BIO_connect_retry(BIO *bio, int timeout)
+                break;
+            }
+        }
+        if (do_retry || timeout < 0) {
```
do_retry && timeout >= 0 ?

DDvO replied:
Oops, yes!
Corrected.

bernd-edlinger commented on crypto/bio/bio_lib.c:
```diff
@@ -786,55 +786,66 @@ void bio_cleanup(void)
- * Parameter timeout == 0 means infinite, < 0 leads to immediate timeout error.
+ * Connect via given BIO using BIO_do_connect() until success/timeout/error.
+ * Parameter timeout == 0 means no timeout, < 0 means exactly one try.
+ * For non-blocking and potentially even non-socket BIOs perform polling with
```
which non-socket BIOs have you in mind?
have you tested this code with one of those?

DDvO replied:
I don't have such a non-socket BIO at hand, so could not test this.
That's part of the reason I wrote "potentially even non-socket BIOs".
Anyway, for non-blocking socket-based BIOs (i.e., timeout > 0) polling is needed.

bernd-edlinger replied:
Well, that's why I'd initially restrict the scope of such a function,
to the domain where it is used, and extend that later
when other non-socket use become available.
I have the impression that must be invented first.

DDvO commented:
Thanks a lot @bernd-edlinger for diligently working with me through the rather fresh high-level BIO connect code. This was much more effort that I would have expected, but certainly worth while because now it is corrected and much cleaner, and also the documentation has improved.

We also fixed some more or less related issues in bss_acpt.c and bss_conn.c.

I've rebased and squashed this PR as much as it made sense.

bernd-edlinger commented:
suggestions for further improvement:
use poll(2) if available,
make whole api depend on OPENSSL_NO_SOCK
unless other concrete use cases can be found.
thus reduce code size.

DDvO commented:
> This looks good to me, but I'll defer to @bernd-edlinger to actually approve

@levitte, now since @bernd-edlinger approved,
 I think this also needs your (formal) blessing.

DDvO commented:
Also here I still need an OTC approval.

This PR should fix by the way #11449.

DDvO commented:
@levitte, is the state of this PR okay for you as well?

openssl-machine commented:
This pull request is ready to merge

DDvO commented:
Merged - thanks again!
